File: leetcode/0118-pascals-triangle/solution.cpp
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<int>> generate(int numRows) {
        vector<vector<int>> triangle;
        if (numRows == 0) return triangle;
        
        triangle.push_back(vector<int>(1, 1));

        for (int RowNum = 1; RowNum < numRows; RowNum++) {
            vector<int> row(RowNum + 1, 1);
            const vector<int>& PreRow = triangle[RowNum - 1];

            for (int j = 1; j < RowNum; j++) {
                row[j] = PreRow[j - 1] + PreRow[j];
            }

            triangle.push_back(row);
        }
        return triangle;
    }
};

File: leetcode/0118-pascals-triangle/solution.java
import java.util.*;
class Solution {
    public List<List<Integer>> generate(int numRows) {
        
        List<List<Integer>> result = new ArrayList<>();
        
        result.add(Arrays.asList(1));
        
        for(int r=2; r<=numRows; r++){
            List<Integer> prev = result.get(result.size()-1);
            List<Integer> curr = new ArrayList<>();
            
            curr.add(1);
            
            for(int j=1; j<r-1; j++){
                
                curr.add(prev.get(j-1)+prev.get(j));
            }
            
            curr.add(1);
            result.add(curr);
        }
        return result;
    }
}
